// @argonjs/argon/src/context.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { autoinject } from 'aurelia-dependency-injection';
import { defined, Entity, Cartographic, DynamicPositionProperty, DynamicProperty, Cartesian3, Quaternion, Matrix3, Matrix4, CesiumMath, Transforms, JulianDate, ReferenceFrame, PerspectiveFrustum } from './cesium/cesium-imports';
import { DEFAULT_NEAR_PLANE, DEFAULT_FAR_PLANE, SerializedSubviewList, SubviewType, CanvasViewport, Viewport, Role } from './common';
import { SessionService } from './session';
import { Event, stringIdentifierFromReferenceFrame, getReachableAncestorReferenceFrames, getSerializedEntityState, getEntityPositionInReferenceFrame, getEntityOrientationInReferenceFrame, deprecated, eastUpSouthToFixedFrame, convertEntityReferenceFrame } from './utils';
import { EntityService, EntityServiceProvider, PoseStatus } from './entity';
import { DeviceService, Device } from './device';
import { ViewService } from './view';
import { PermissionServiceProvider, PermissionState } from './permission';
import { RealityService, RealityServiceProvider } from './reality';
/**
 * Provides a means of querying the current state of reality.
 */
var ContextService = (function () {
    function ContextService(entityService, sessionService, deviceService, viewService, realityService) {
        var _this = this;
        this.entityService = entityService;
        this.sessionService = sessionService;
        this.deviceService = deviceService;
        this.viewService = viewService;
        this.realityService = realityService;
        /**
         * An event that is raised after managed entities have been updated for
         * the current frame.
         */
        this.updateEvent = new Event();
        /**
         * An event that is raised when it is an approriate time to render graphics.
         * This event fires after the update event.
         */
        this.renderEvent = new Event();
        /**
         * An event that is raised after the render event
         */
        this.postRenderEvent = new Event();
        /**
         * An event that fires when the origin changes.
         */
        this.originChangeEvent = new Event();
        /**
         * A monotonically increasing value (in milliseconds) for the current frame state.
         * This value is useful only for doing accurate *timing*, not for determining
         * the absolute time. Use [[ContextService.time]] for absolute time.
         * This value is -1 until the first [[ContextService.updateEvent]].
         */
        this.timestamp = -1;
        /**
         * The time in milliseconds since the previous timestamp,
         * capped to [[ContextService.maxDeltaTime]]
         */
        this.deltaTime = 0;
        /**
         * This value caps the deltaTime for each frame. By default,
         * the value is 1/3s (333.3ms)
         */
        this.maxDeltaTime = 1 / 3 * 1000;
        /**
         * The current (absolute) time according to the current reality.
         * This value is arbitrary until the first [[ContextService.updateEvent]].
         */
        this.time = new JulianDate(0, 0);
        /**
        * An entity representing the origin, which is oriented
        * with +Y up. The origin changes infrequently, is platform dependent,
        * and is the suggested origin for a rendering scenegraph.
        *
        * Any time the origin changes, the originChange event is raised.
        */
        this.origin = this.entities.add(new Entity({
            id: 'ar.origin',
            name: 'Origin',
            position: new DynamicPositionProperty(undefined, ReferenceFrame.FIXED),
            orientation: new DynamicProperty(undefined)
        }));
        this._localOrigin = this.entities.add(new Entity({
            id: 'ar.localOrigin',
            name: 'Local Origin',
            position: new DynamicPositionProperty(Cartesian3.ZERO, this.origin),
            orientation: new DynamicProperty(Quaternion.IDENTITY)
        }));
        this._localOriginEastNorthUp = this.entities.add(new Entity({
            id: 'ar.localOriginENU',
            name: 'Local Origin (ENU)',
            position: new DynamicPositionProperty(Cartesian3.ZERO, this.origin),
            orientation: new DynamicProperty(Quaternion.fromAxisAngle(Cartesian3.UNIT_X, -Math.PI / 2))
        }));
        /**
         * A coordinate system representing the physical space in which the user is free to move
         * around with high-precision tracking, and positioned on the surface the user is standing on.
         * This space is oriented such that a platform-dependent "forward" corresponds to -Z.
         */
        this.stage = this.entities.add(new Entity({
            id: 'ar.stage',
            name: 'Stage',
            position: new DynamicPositionProperty(undefined, ReferenceFrame.FIXED),
            orientation: new DynamicProperty(undefined)
        }));
        /**
         * A coordinate system positioned at the stage,
         * where +X is east, +Y is up, and +Z is south (East-Up-South), if geolocation is known.
         * If geolocation is unknown, this entity has an undefined pose.
         */
        this.stageEUS = this.entities.add(new Entity({
            id: 'ar.stageEUS',
            name: 'Stage (EUS)',
            position: new DynamicPositionProperty(undefined, this.stage),
            orientation: new DynamicProperty(undefined)
        }));
        /**
         * A coordinate system positioned at the stage,
         * where +X is east, +Y is north, and +Z is up (East-North-Up), if geolocation is known.
         * If geolocation is unknown, this entity has an undefined pose.
         */
        this.stageENU = this.entities.add(new Entity({
            id: 'ar.stageENU',
            name: 'Stage (ENU)',
            position: new DynamicPositionProperty(undefined, this.stage),
            orientation: new DynamicProperty(undefined)
        }));
        /**
         * A coordinate system representing the floor.
         * While the `stage` always represents a physical surface,
         * the `floor` entity may represent a virtual floor.
         */
        this.floor = this.entities.add(new Entity({
            id: 'ar.floor',
            name: 'Floor',
            position: new DynamicPositionProperty(Cartesian3.ZERO, this.stage),
            orientation: new DynamicProperty(Quaternion.IDENTITY)
        }));
        /**
         * An coordinate system representing the user,
         * where +X is right, +Y is up, and -Z is the direction the user is facing
         */
        this.user = this.entities.add(new Entity({
            id: 'ar.user',
            name: 'User',
            position: new DynamicPositionProperty(undefined, this.stage),
            orientation: new DynamicProperty(undefined)
        }));
        /**
         * An coordinate system representing the rendering view,
         * where +X is right, +Y is up, and -Z is the direction of the view.
         */
        this.view = this.entities.add(new Entity({
            id: 'ar.view',
            name: 'View',
            position: new DynamicPositionProperty(Cartesian3.ZERO, this.user),
            orientation: new DynamicProperty(Quaternion.IDENTITY)
        }));
        /**
         * The default reference frame to use when calling `getEntityPose`.
         * By default, this is the `origin` reference frame.
         */
        this.defaultReferenceFrame = this.origin;
        this._updatingEntities = new Set();
        this._knownEntities = new Set();
        this._scratchCartesian = new Cartesian3;
        this._scratchQuaternion = new Quaternion;
        this._scratchFrustum = new PerspectiveFrustum();
        /**
         * Subscribe to pose updates for the given entity id
         *
         * @returns A Promise that resolves to a new or existing entity
         * instance matching the given id, if the subscription is successful
         */
        this.subscribe = this.entityService.subscribe.bind(this.entityService);
        /**
         * Unsubscribe to pose updates for the given entity id
         */
        this.unsubscribe = this.entityService.unsubscribe.bind(this.entityService);
        this._frameIndex = -1;
        this._scratchFrameState = {
            time: {},
            entities: {},
            viewport: {},
            subviews: []
        };
        this._getSerializedEntityState = getSerializedEntityState;
        this._getEntityPositionInReferenceFrame = getEntityPositionInReferenceFrame;
        this._getEntityOrientationInReferenceFrame = getEntityOrientationInReferenceFrame;
        this._eastUpSouthToFixedFrame = eastUpSouthToFixedFrame;
        this._eastNorthUpToFixedFrame = Transforms.eastNorthUpToFixedFrame;
        this._getReachableAncestorReferenceFrames = getReachableAncestorReferenceFrames;
        this._scratchArray = [];
        this._scratchMatrix3 = new Matrix3;
        this._scratchMatrix4 = new Matrix4;
        this._convertEntityReferenceFrame = convertEntityReferenceFrame;
        this.sessionService.manager.on['ar.context.update'] = function (state) {
            var scratchFrustum = _this._scratchFrustum;
            // backwards-compat
            if (typeof state.reality !== 'string') {
                state.reality = state.reality && state.reality['uri'];
            }
            if (!state.viewport && state['view'] && state['view'].viewport) {
                state.viewport = state['view'].viewport;
            }
            if (!state.subviews && state['view'] && state['view'].subviews) {
                state.subviews = state['view'].subviews;
                scratchFrustum.near = DEFAULT_NEAR_PLANE;
                scratchFrustum.far = DEFAULT_FAR_PLANE;
                for (var _i = 0, _a = state.subviews; _i < _a.length; _i++) {
                    var s = _a[_i];
                    var frustum = s['frustum'];
                    scratchFrustum.xOffset = frustum.xOffset || 0;
                    scratchFrustum.yOffset = frustum.yOffset || 0;
                    scratchFrustum.fov = frustum.fov || CesiumMath.PI_OVER_THREE;
                    scratchFrustum.aspectRatio = frustum.aspectRatio || 1;
                    s.projectionMatrix = Matrix4.clone(scratchFrustum.projectionMatrix, s.projectionMatrix);
                }
            }
            if (!state.entities[_this.user.id] && state['view'] && state['view'].pose) {
                state.entities[_this.user.id] = state['view'].pose;
            }
            // end backwards-compat
            // the `skipEvents` flag skips update/render events,
            // allowing a reality to process 'ar.context.update' so it knows 
            // what the current state of the system is, while maintaining control
            // over it's frame timing (by calling submitFrameState)
            var skipEvents = _this.sessionService.isRealityViewer;
            _this._update(state, skipEvents);
        };
        this._scratchFrustum.near = DEFAULT_NEAR_PLANE;
        this._scratchFrustum.far = DEFAULT_FAR_PLANE;
        this._scratchFrustum.fov = CesiumMath.PI_OVER_THREE;
        this._scratchFrustum.aspectRatio = 1;
        this._serializedFrameState = {
            reality: undefined,
            time: JulianDate.now(),
            entities: {},
            viewport: new CanvasViewport,
            subviews: [{
                    type: SubviewType.SINGULAR,
                    viewport: new Viewport,
                    projectionMatrix: this._scratchFrustum.projectionMatrix
                }],
        };
    }
    Object.defineProperty(ContextService.prototype, "entities", {
        get: function () { return this.entityService.collection; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "localOriginChangeEvent", {
        /**
         * An event that fires when the local origin changes.
         */
        get: function () { return this.originChangeEvent; },
        enumerable: true,
        configurable: true
    });
    ;
    Object.defineProperty(ContextService.prototype, "userTracking", {
        /**
         * Returns the DOF support of the reality.
         * "none"|"3DOF"|"6DOF"
         */
        get: function () {
            return this._serializedFrameState && this._serializedFrameState.userTracking || '3DOF';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "localOrigin", {
        /** alias for origin */
        get: function () { return this._localOrigin; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "localOriginEastUpSouth", {
        // To be removed. This is no longer useful.
        get: function () { return this._localOrigin; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "localOriginEastNorthUp", {
        // To be removed. This is no longer useful.
        get: function () { return this._localOriginEastNorthUp; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "serializedFrameState", {
        /**
         * The serialized frame state for this frame
         */
        get: function () {
            return this._serializedFrameState;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "systemTime", {
        /**
         * Deprecated. Use timestamp property.
         * @private
         */
        get: function () {
            return this.timestamp;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Deprecated. To be removed.
     * @private
     */
    ContextService.prototype.getTime = function () {
        return this.time;
    };
    /**
     * Deprecated. To be removed. Use the defaultReferenceFrame property if necessary.
     * @private
     */
    ContextService.prototype.setDefaultReferenceFrame = function (origin) {
        this.defaultReferenceFrame = origin;
    };
    /**
     * Deprecated. To be removed.  Use the defaultReferenceFrame property.
     * @private
     */
    ContextService.prototype.getDefaultReferenceFrame = function () {
        return this.defaultReferenceFrame;
    };
    /**
     * Subscribe to pose updates for an entity specified by the given id
     *
     * @deprecated Use [[ContextService#subscribe]]
     * @param id - the id of the desired entity
     * @returns A new or existing entity instance matching the given id
     */
    ContextService.prototype.subscribeToEntityById = function (id) {
        this.subscribe(id);
        return this.entities.getOrCreateEntity(id);
    };
    /**
     * Get the cartographic position of an Entity for the current context time
     */
    ContextService.prototype.getEntityCartographic = function (entity, result) {
        return this.entityService.getCartographic(entity, this.time, result);
    };
    /**
     * Deprecated. Use `EntityService.createFixed` (`app.entity.createFixed`);
     */
    ContextService.prototype.createGeoEntity = function (cartographic, localToFixed) {
        return this.entityService.createFixed(cartographic, localToFixed);
    };
    /**
     * Create a new EntityPose instance to represent the pose of an entity
     * relative to a given reference frame. If no reference frame is specified,
     * then the pose is based on the context's defaultReferenceFrame.
     *
     * @param entityOrId - the entity to track
     * @param referenceFrameOrId - The intended reference frame. Defaults to `this.defaultReferenceFrame`.
     */
    ContextService.prototype.createEntityPose = function (entityOrId, referenceFrameOrId) {
        if (referenceFrameOrId === void 0) { referenceFrameOrId = this.defaultReferenceFrame; }
        return this.entityService.createEntityPose(entityOrId, referenceFrameOrId);
    };
    /**
     * Gets the current pose of an entity, relative to a given reference frame.
     *
     * @param entityOrId - The entity whose state is to be queried.
     * @param referenceFrameOrId - The intended reference frame. Defaults to `this.defaultReferenceFrame`.
     */
    ContextService.prototype.getEntityPose = function (entityOrId, referenceFrameOrId) {
        if (referenceFrameOrId === void 0) { referenceFrameOrId = this.defaultReferenceFrame; }
        return this.entityService.getEntityPose(entityOrId, referenceFrameOrId, this.time);
    };
    /**
     * Process the next frame state (which should come from the current reality viewer)
     */
    ContextService.prototype.submitFrameState = function (frameState) {
        frameState.index = ++this._frameIndex;
        this._update(frameState);
    };
    /**
     * Create a frame state.
     *
     * @param time
     * @param viewport
     * @param subviewList
     * @param userTracking
     * @param entityOptions
     */
    ContextService.prototype.createFrameState = function (time, viewport, subviewList, options) {
        var overrideUser = options && options.overrideUser;
        if (this.deviceService.strict) {
            if (overrideUser) {
                overrideUser = false;
            }
        }
        var frameState = this._scratchFrameState;
        frameState.time = JulianDate.clone(time, frameState.time);
        frameState.viewport = CanvasViewport.clone(viewport, frameState.viewport);
        frameState.subviews = SerializedSubviewList.clone(subviewList, frameState.subviews);
        var entities = frameState.entities = {};
        var getSerializedEntityState = this._getSerializedEntityState;
        // stage
        var stage = this.stage;
        if (options && options.overrideStage) {
            entities[stage.id] = getSerializedEntityState(stage, time, undefined);
        }
        // user
        var user = this.user;
        if (overrideUser) {
            entities[user.id] = getSerializedEntityState(user, time, stage);
        }
        // view
        var view = this.view;
        if (options && options.overrideView) {
            entities[view.id] = getSerializedEntityState(view, time, user);
        }
        // subviews
        if (options && options.overrideSubviews) {
            for (var index = 0; index < subviewList.length; index++) {
                // check for valid projection matrices
                var subview = subviewList[index];
                if (!isFinite(subview.projectionMatrix[0]))
                    throw new Error('Invalid projection matrix (contains non-finite values)');
                // subview
                var subviewEntity = this.getSubviewEntity(index);
                entities[subviewEntity.id] = getSerializedEntityState(subviewEntity, time, view);
            }
        }
        // floor
        var floorOffset = options && options.floorOffset || 0;
        var floor = this.floor;
        floor.position.setValue(Cartesian3.fromElements(0, floorOffset, 0, this._scratchCartesian), stage);
        if (floorOffset !== 0) {
            frameState.entities[this.floor.id] = getSerializedEntityState(floor, time, stage);
        }
        // user tracking
        frameState.userTracking = options && options.userTracking;
        return frameState;
    };
    // All of the following work is only necessary when running in an old manager (version === 0)
    ContextService.prototype._updateBackwardsCompatability = function (frameState) {
        this._knownEntities.clear();
        // update the entities the manager knows about
        var entityService = this.entityService;
        for (var id in frameState.entities) {
            entityService.updateEntityFromSerializedState(id, frameState.entities[id]);
            this._updatingEntities.add(id);
            this._knownEntities.add(id);
        }
        // if the mangager didn't send us an update for a particular entity,
        // assume the manager no longer knows about it
        for (var _i = 0, _a = this._updatingEntities; _i < _a.length; _i++) {
            var id = _a[_i];
            if (!this._knownEntities.has(id)) {
                var entity = this.entities.getById(id);
                if (entity) {
                    if (entity.position)
                        entity.position.setValue(undefined);
                    if (entity.orientation)
                        entity.orientation.setValue(undefined);
                }
                this._updatingEntities.delete(id);
            }
        }
        // If running within an older manager, we have to set the stage based on the user pose. 
        var userPositionFixed = this._getEntityPositionInReferenceFrame(this.user, frameState.time, ReferenceFrame.FIXED, this._scratchCartesian);
        if (userPositionFixed) {
            var eusToFixedFrameTransform = eastUpSouthToFixedFrame(userPositionFixed, undefined, this._scratchMatrix4);
            var eusRotationMatrix = Matrix4.getRotation(eusToFixedFrameTransform, this._scratchMatrix3);
            var eusOrientation = Quaternion.fromRotationMatrix(eusRotationMatrix);
            this.stage.position.setValue(userPositionFixed, ReferenceFrame.FIXED);
            this.stage.orientation.setValue(eusOrientation);
        }
        else {
            this.stage.position.setValue(Cartesian3.fromElements(0, -this.deviceService.suggestedUserHeight, 0, this._scratchCartesian), this.user.position.referenceFrame);
            this.stage.orientation.setValue(Quaternion.IDENTITY);
        }
        frameState.entities[this.stage.id] = true; // assume overriden for _update
    };
    // TODO: This function is called a lot. Potential for optimization. 
    ContextService.prototype._update = function (frameState, skipEvents) {
        this._serializedFrameState = frameState;
        // update our time values
        var timestamp = performance.now();
        this.deltaTime = Math.min(timestamp - this.timestamp, this.maxDeltaTime);
        this.timestamp = timestamp;
        JulianDate.clone(frameState.time, this.time);
        // update provided entities
        if (this.sessionService.manager.isConnected && this.sessionService.manager.version[0] === 0) {
            this._updateBackwardsCompatability(frameState);
        }
        else {
            this._updateEntities(frameState.entities);
        }
        // update device entities (device service needs to know if manager is overriding it's entities)
        this.deviceService._processContextFrameState(frameState);
        // update context entities
        this._updateContextEntities(frameState);
        this._updateStageGeo();
        // update view and reality service
        this.viewService._processContextFrameState(frameState, this);
        this.realityService._processContextFrameState(frameState);
        // exit early if onlyUpdate flag is set
        if (skipEvents)
            return;
        // raise origin change event if necessary
        this._checkOriginChange();
        // raise events for the user to update and render the scene
        this.updateEvent.raiseEvent(this);
        this.renderEvent.raiseEvent(this);
        this.postRenderEvent.raiseEvent(this);
        // publish frameState to the manager (noop if this session is not a reality)
        this.realityService._publishContextFrameState(frameState);
        // submit frame if necessary
        this._trySubmitFrame();
    };
    ContextService.prototype._updateEntities = function (entities) {
        var entityService = this.entityService;
        for (var id in entities) {
            entityService.updateEntityFromSerializedState(id, entities[id]);
        }
    };
    ContextService.prototype._updateContextEntities = function (frameState) {
        var time = frameState.time;
        var entities = frameState.entities;
        // If the reality did not override the stage, then set the origin to the device origin.
        // Otherwise, find the appropriate origin pose. 
        var contextOrigin = this.origin;
        var contextStage = this.stage;
        var deviceOrigin = this.deviceService.origin;
        var deviceStage = this.deviceService.stage;
        var contextStagePosition = contextStage.position;
        var contextStageOrientation = contextStage.orientation;
        var contextOriginPosition = contextOrigin.position;
        var contextOriginOrientation = contextOrigin.orientation;
        if (!entities[contextStage.id]) {
            contextStagePosition.setValue(Cartesian3.ZERO, deviceStage);
            contextStageOrientation.setValue(Quaternion.IDENTITY);
            contextOriginPosition.setValue(Cartesian3.ZERO, deviceOrigin);
            contextOriginOrientation.setValue(Quaternion.IDENTITY);
        }
        else {
            // save stage relative to origin
            var stageRelativeToOrigin = this.getEntityPose(deviceStage, deviceOrigin);
            // first figure out where the origin should be
            var originRelativeToStage = this.getEntityPose(deviceOrigin, deviceStage);
            contextOriginPosition.setValue(originRelativeToStage.position, contextStage);
            contextOriginOrientation.setValue(originRelativeToStage.orientation);
            // convert origin to the same frame used by the overridden stage
            var rootFrame = getReachableAncestorReferenceFrames(contextStage, time, this._scratchArray)[0];
            if (!defined(rootFrame))
                throw new Error("Stage frame must have a reachable parent reference frame!");
            var originRelativeToRoot = this.getEntityPose(contextOrigin, rootFrame);
            contextOriginPosition.setValue(originRelativeToRoot.position, rootFrame);
            contextOriginOrientation.setValue(originRelativeToRoot.orientation);
            // convert stage to be relative to origin
            contextStagePosition.setValue(stageRelativeToOrigin.position, contextOrigin);
            contextStageOrientation.setValue(stageRelativeToOrigin.orientation);
        }
        // update user entity (relative to origin) based on device user (relative to device origin) if the reality did not override it
        var deviceUser = this.deviceService.user;
        var contextUser = this.user;
        if (!entities[contextUser.id]) {
            var userRelativeToOrigin = this.getEntityPose(deviceUser, deviceOrigin);
            var contextUserPosition = contextUser.position;
            var contextUserOrientation = contextUser.orientation;
            contextUserPosition.setValue(userRelativeToOrigin.position, contextOrigin);
            contextUserOrientation.setValue(userRelativeToOrigin.orientation);
        }
        else {
            this._convertEntityReferenceFrame(contextUser, time, contextStage);
        }
        // update view entity (if the reality did not set it)
        var contextView = this.view;
        if (!entities[contextView.id]) {
            var contextViewPosition = contextView.position;
            var contextViewOrientation = contextView.orientation;
            contextViewPosition.setValue(Cartesian3.ZERO, contextUser);
            contextViewOrientation.setValue(Quaternion.IDENTITY);
        }
        else {
            this._convertEntityReferenceFrame(contextView, time, contextUser);
        }
        // update subview entities (if the reality did not set them)
        for (var i = 0; i < frameState.subviews.length; i++) {
            if (!entities['ar.view_' + i]) {
                var deviceSubview = this.deviceService.getSubviewEntity(i);
                var contextSubview = this.getSubviewEntity(i);
                var subviewPositionValue = this._getEntityPositionInReferenceFrame(deviceSubview, time, deviceUser, this._scratchCartesian);
                var subviewOrientationValue = this._getEntityOrientationInReferenceFrame(deviceSubview, time, deviceUser, this._scratchQuaternion);
                var contextSubviewPosition = contextSubview.position;
                var contextSubviewOrientation = contextSubview.orientation;
                contextSubviewPosition.setValue(subviewPositionValue, contextView);
                contextSubviewOrientation.setValue(subviewOrientationValue);
            }
        }
        // update floor entity (if the reality did not set it)
        if (!entities[this.floor.id]) {
            var floorPosition = this.floor.position;
            floorPosition.setValue(Cartesian3.ZERO, contextStage);
        }
        else {
            this._convertEntityReferenceFrame(this.floor, time, contextStage);
        }
    };
    ContextService.prototype._updateStageGeo = function () {
        // update stageEUS and stageENU
        var time = this.time;
        var contextStage = this.stage;
        var stageFixedPosition = this._getEntityPositionInReferenceFrame(contextStage, time, ReferenceFrame.FIXED, this._scratchCartesian);
        if (stageFixedPosition) {
            // EUS
            var eusTransform = this._eastUpSouthToFixedFrame(stageFixedPosition, undefined, this._scratchMatrix4);
            var eusRotation = Matrix4.getRotation(eusTransform, this._scratchMatrix3);
            var eusOrientation = Quaternion.fromRotationMatrix(eusRotation, this._scratchQuaternion);
            this.stageEUS.position.setValue(stageFixedPosition, ReferenceFrame.FIXED);
            this.stageEUS.orientation.setValue(eusOrientation);
            // ENU
            var enuTransform = this._eastNorthUpToFixedFrame(stageFixedPosition, undefined, this._scratchMatrix4);
            var enuRotation = Matrix4.getRotation(enuTransform, this._scratchMatrix3);
            var enuOrientation = Quaternion.fromRotationMatrix(enuRotation, this._scratchQuaternion);
            this.stageENU.position.setValue(stageFixedPosition, ReferenceFrame.FIXED);
            this.stageENU.orientation.setValue(enuOrientation);
        }
        else {
            this.stageEUS.position.setValue(undefined, ReferenceFrame.FIXED);
            this.stageEUS.orientation.setValue(undefined);
            this.stageENU.position.setValue(undefined, ReferenceFrame.FIXED);
            this.stageENU.orientation.setValue(undefined);
        }
    };
    ContextService.prototype._checkOriginChange = function () {
        var time = this.time;
        var originReferenceFrame = this._getReachableAncestorReferenceFrames(this.origin, time, this._scratchArray)[0] || ReferenceFrame.FIXED;
        var originPose = this.getEntityPose(this.origin, originReferenceFrame);
        if (originReferenceFrame !== this._previousOriginReferenceFrame || originPose.status & PoseStatus.CHANGED) {
            this._previousOriginReferenceFrame = originReferenceFrame;
            if (this.sessionService.isRealityAugmenter)
                console.log('Updated context origin to ' + JSON.stringify(originPose.position) + " at " + stringIdentifierFromReferenceFrame(originReferenceFrame));
            this.originChangeEvent.raiseEvent(undefined);
        }
    };
    ContextService.prototype._trySubmitFrame = function () {
        var vrDisplay = this.deviceService.vrDisplay;
        if (this.deviceService.autoSubmitFrame &&
            vrDisplay && vrDisplay.isPresenting &&
            !this.sessionService.isRealityViewer) {
            vrDisplay.submitFrame();
        }
    };
    ContextService.prototype.getSubviewEntity = function (index) {
        var subviewEntity = this.entityService.collection.getOrCreateEntity('ar.view_' + index);
        if (!subviewEntity.position) {
            subviewEntity.position = new DynamicPositionProperty(Cartesian3.ZERO, this.user);
        }
        if (!subviewEntity.orientation) {
            subviewEntity.orientation = new DynamicProperty(Quaternion.IDENTITY);
        }
        return subviewEntity;
    };
    ContextService.prototype.subscribeGeolocation = function (options) {
        var _this = this;
        return this.sessionService.manager.whenConnected().then(function () {
            if (_this.sessionService.manager.versionNumber >= 1.4)
                return _this.entityService.subscribe(_this.origin.id, options).then(function () { });
            else
                return _this.entityService.subscribe(_this.stage.id, options).then(function () { });
        });
    };
    ContextService.prototype.unsubscribeGeolocation = function () {
        var _this = this;
        this.sessionService.manager.whenConnected().then(function () {
            if (_this.sessionService.manager.versionNumber >= 1.4)
                _this.entityService.unsubscribe(_this.origin.id);
            else
                _this.entityService.unsubscribe(_this.stage.id);
        });
    };
    ContextService.prototype._getInheritedMetaProperty = function (entity, metaPropKey) {
        var referenceFrame = entity.position.referenceFrame;
        while (referenceFrame) {
            var metaProp = referenceFrame['meta'] && referenceFrame['meta'][metaPropKey];
            if (defined(metaProp))
                return metaProp;
            referenceFrame = referenceFrame.position && referenceFrame.position.referenceFrame;
        }
    };
    Object.defineProperty(ContextService.prototype, "geoHeadingAccuracy", {
        /**
         * If geopose is available, this is the accuracy of the user heading
         */
        get: function () {
            return this._getInheritedMetaProperty(this.user, 'geoHeadingAccuracy');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "geoHorizontalAccuracy", {
        /**
         * If geopose is available, this is the horizontal accuracy of the user geolocation
         */
        get: function () {
            return this._getInheritedMetaProperty(this.stage, 'geoHorizontalAccuracy');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "geoVerticalAccuracy", {
        /**
         * If geopose is available, this is the vertical accuracy of the user geolocation
         */
        get: function () {
            return this._getInheritedMetaProperty(this.stage, 'geoVerticalAccuracy');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "geoposeHeadingAccuracy", {
        /**
         * @deprecated
         */
        get: function () {
            return this.geoHeadingAccuracy;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "geoposeHorizontalAccuracy", {
        /**
         * @deprecated
         */
        get: function () {
            return this.geoHorizontalAccuracy;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextService.prototype, "geoposeVerticalAccuracy", {
        /**
         * @deprecated
         */
        get: function () {
            return this.geoVerticalAccuracy;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        deprecated('originChangeEvent'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], ContextService.prototype, "localOriginChangeEvent", null);
    __decorate([
        deprecated('origin'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], ContextService.prototype, "localOrigin", null);
    __decorate([
        deprecated(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], ContextService.prototype, "localOriginEastUpSouth", null);
    __decorate([
        deprecated(),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], ContextService.prototype, "localOriginEastNorthUp", null);
    __decorate([
        deprecated('timestamp'),
        __metadata("design:type", Object),
        __metadata("design:paramtypes", [])
    ], ContextService.prototype, "systemTime", null);
    __decorate([
        deprecated('time'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", JulianDate)
    ], ContextService.prototype, "getTime", null);
    __decorate([
        deprecated(),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Object]),
        __metadata("design:returntype", void 0)
    ], ContextService.prototype, "setDefaultReferenceFrame", null);
    __decorate([
        deprecated('defaultReferenceFrame'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", []),
        __metadata("design:returntype", Object)
    ], ContextService.prototype, "getDefaultReferenceFrame", null);
    __decorate([
        deprecated('subscribe'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [String]),
        __metadata("design:returntype", Entity)
    ], ContextService.prototype, "subscribeToEntityById", null);
    __decorate([
        deprecated('EntityService.createFixed'),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [Cartographic, Object]),
        __metadata("design:returntype", void 0)
    ], ContextService.prototype, "createGeoEntity", null);
    ContextService = __decorate([
        autoinject(),
        __metadata("design:paramtypes", [EntityService,
            SessionService,
            DeviceService,
            ViewService,
            RealityService])
    ], ContextService);
    return ContextService;
}());
export { ContextService };
var ContextServiceProvider = (function () {
    function ContextServiceProvider(sessionService, contextService, deviceService, entityServiceProvider, permissionServiceProvider, realityServiceProvider, device) {
        var _this = this;
        this.sessionService = sessionService;
        this.contextService = contextService;
        this.deviceService = deviceService;
        this.entityServiceProvider = entityServiceProvider;
        this.permissionServiceProvider = permissionServiceProvider;
        this.realityServiceProvider = realityServiceProvider;
        this.device = device;
        this._cacheTime = new JulianDate(0, 0);
        this._sessionEntities = {};
        // private _temp:any = {};
        this._includedFrames = {};
        this._excludedFrames = {};
        this.desiredGeolocationOptions = {};
        this.sessionGeolocationOptions = new Map();
        // subscribe to context geolocation if any child sessions have subscribed
        this.entityServiceProvider.sessionSubscribedEvent.addEventListener(function (evt) {
            if (evt.id === _this.contextService.stage.id && evt.session !== _this.sessionService.manager) {
                _this._setGeolocationOptions(evt.session, evt.options);
                _this.contextService.subscribeGeolocation(_this.desiredGeolocationOptions);
            }
        });
        // unsubscribe from context geolocation if all child sessions are unsubscribed
        this.entityServiceProvider.sessionUnsubscribedEvent.addEventListener(function () {
            var subscribers = _this.entityServiceProvider.subscribersByEntity.get(_this.contextService.stage.id);
            if (subscribers && subscribers.size === 1 && subscribers.has(_this.sessionService.manager)) {
                _this.contextService.unsubscribeGeolocation();
            }
        });
        this.sessionService.connectEvent.addEventListener(function (session) {
            var subscriptions = _this.entityServiceProvider.subscriptionsBySubscriber.get(session);
            subscriptions[_this.contextService.user.id] = {};
            subscriptions[_this.contextService.stage.id] = {};
            subscriptions[_this.deviceService.user.id] = {};
            subscriptions[_this.deviceService.stage.id] = {};
        });
        // submit frame state from reality
        this.realityServiceProvider.nextFrameStateEvent.addEventListener(function (frameState) {
            _this.contextService.submitFrameState(frameState);
            _this._publishFrameState(); // publish frame state to child sessions
        });
    }
    ContextServiceProvider.prototype._publishFrameState = function () {
        var state = this.contextService.serializedFrameState;
        this._cacheTime = JulianDate.clone(state.time, this._cacheTime);
        for (var _i = 0, _a = this.sessionService.managedSessions; _i < _a.length; _i++) {
            var session = _a[_i];
            this._sendUpdateForSession(state, session);
        }
    };
    ContextServiceProvider.prototype._sendUpdateForSession = function (state, session) {
        if (Role.isRealityViewer(session.info.role) && session.versionNumber < 1.4) {
            return; // older reality versions don't expect to receive update events from manager
        }
        var sessionEntities = this._sessionEntities;
        var entityServiceProvider = this.entityServiceProvider;
        // clear session entities
        for (var id in sessionEntities) {
            delete sessionEntities[id];
        }
        // reference all entities from the primary frame state.
        if (state.entities) {
            for (var id in state.entities) {
                sessionEntities[id] = state.entities[id];
            }
        }
        // identity frames to provide to the session
        var includedFrames = this._includedFrames;
        for (id in includedFrames)
            delete includedFrames[id]; //clear
        var subscriptions = entityServiceProvider.subscriptionsBySubscriber.get(session);
        if (subscriptions) {
            for (var id in subscriptions) {
                includedFrames[id] = true;
            }
        }
        var deviceService = this.deviceService;
        var deviceOriginId = deviceService.origin.id;
        var deviceStageId = deviceService.stage.id;
        var deviceUserId = deviceService.stage.id;
        var contextService = this.contextService;
        var contextOriginId = contextService.origin.id;
        var contextStageId = contextService.stage.id;
        var contextUserId = contextService.stage.id;
        includedFrames[deviceOriginId] = true;
        includedFrames[deviceStageId] = true;
        includedFrames[deviceUserId] = true;
        includedFrames[contextOriginId] = true;
        includedFrames[contextStageId] = true;
        includedFrames[contextUserId] = true;
        includedFrames[contextService.view.id] = true;
        for (var i = 0; i < state.subviews.length; i++) {
            includedFrames['ar.view_' + i] = true;
            includedFrames['ar.device.view_' + i] = true;
        }
        // identify frames to hide from the session
        var excludedFrames = this._excludedFrames;
        for (id in excludedFrames)
            delete excludedFrames[id]; //clear
        // exclude geolocated frames if necessary 
        if (this.permissionServiceProvider.getPermissionState(session, 'geolocation') != PermissionState.GRANTED) {
            excludedFrames[deviceOriginId] = true;
            excludedFrames[contextOriginId] = true;
        }
        // get states for all included frames, minus excluded frames
        entityServiceProvider.fillEntityStateMap(sessionEntities, state.time, includedFrames, excludedFrames);
        // remove device orientation frame since each session can get this directly
        if (session.versionNumber >= 1.4)
            delete sessionEntities[this.device.deviceOrientation.id];
        // recycle the frame state object, but with the session entities
        var parentEntities = state.entities;
        state.entities = sessionEntities;
        state.time = state.time;
        state.sendTime = JulianDate.now(state.sendTime);
        // send
        session.send('ar.context.update', state);
        // restore the parent entities
        state.entities = parentEntities;
    };
    ContextServiceProvider.prototype._setGeolocationOptions = function (session, options) {
        var _this = this;
        this.sessionGeolocationOptions.set(session, options);
        session.closeEvent.addEventListener(function () {
            _this.sessionGeolocationOptions.delete(session);
            _this._updateDesiredGeolocationOptions();
        });
        this._updateDesiredGeolocationOptions();
    };
    ContextServiceProvider.prototype._updateDesiredGeolocationOptions = function () {
        var reducedOptions = {};
        this.sessionGeolocationOptions.forEach(function (options, session) {
            reducedOptions.enableHighAccuracy =
                reducedOptions.enableHighAccuracy || (options && options.enableHighAccuracy) || false;
        });
        if (this.desiredGeolocationOptions.enableHighAccuracy !== reducedOptions.enableHighAccuracy) {
            this.desiredGeolocationOptions = reducedOptions;
        }
    };
    ContextServiceProvider = __decorate([
        autoinject(),
        __metadata("design:paramtypes", [SessionService,
            ContextService,
            DeviceService,
            EntityServiceProvider,
            PermissionServiceProvider,
            RealityServiceProvider,
            Device])
    ], ContextServiceProvider);
    return ContextServiceProvider;
}());
export { ContextServiceProvider };
